File: Reqnroll/Chapter16/16.2-ContextShortcutPattern/After/WIMP.Specs/StepDefinitions/OrderingStepDefinitions.cs
using Reqnroll;

using WIMP.App.Models;
using WIMP.Specs.Drivers;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class OrderingStepDefinitions(OrderingContext orderingContext, BackdoorApiDriver backdoorApiDriver)
{
    [Given("the customer has an order that is waiting for pickup")]
    public async Task GivenTheCustomerHasAnOrderThatIsWaitingForPickup()
    {
        var orderRequest = new PlaceOrderRequestObjectMother().Build();
        var placedOrder = await backdoorApiDriver
            .PrepareOrder(DomainDefaults.CustomerName, orderRequest, OrderStatus.WaitingForPickup)
            .Execute();
        orderingContext.CurrentOrderNo = placedOrder.OrderNo;
    }
}

File: CucumberJVM/Chapter16/16.2-ContextShortcutPattern/After/src/test/java/com/wimp/app/specs/stepdefinitions/OrderingStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.models.OrderStatus;
import com.wimp.app.specs.drivers.*;
import com.wimp.app.specs.support.DomainDefaults;
import com.wimp.app.specs.support.OrderingContext;
import com.wimp.app.specs.support.PlaceOrderRequestObjectMother;
import io.cucumber.java.en.*;

public class OrderingStepDefinitions {
    private final OrderingContext orderingContext;
    private final BackdoorApiDriver backdoorApiDriver;

    public OrderingStepDefinitions(OrderingContext orderingContext, BackdoorApiDriver backdoorApiDriver) {
        this.orderingContext = orderingContext;
        this.backdoorApiDriver = backdoorApiDriver;
    }

    @Given("the customer has an order that is waiting for pickup")
    public void theCustomerHasAnOrderThatIsWaitingForPickup() throws Exception {
        var orderRequest = new PlaceOrderRequestObjectMother().build();
        var placedOrder = backdoorApiDriver
            .prepareOrder(DomainDefaults.CUSTOMER_NAME, orderRequest, OrderStatus.WAITING_FOR_PICKUP)
            .execute();
        orderingContext.setCurrentOrderNo(placedOrder.getOrderNo());
    }
}
